pkg/controller/machineset/machineset_controller.go

```diff
@@ -0,0 +1,191 @@
+func (r *ReconcileMachineSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := log.WithValues("MachineSet", request.NamespacedName.String())
+	ctx := context.TODO()
+	// Fetch the MachineSet instance
+	instance := &machinev1beta1.MachineSet{}
+	err := r.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Don't take action if the annotation is not present
+	annotations := instance.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		return reconcile.Result{}, nil
+	}
+	_, present := annotations[AutoScaleAnnotation]
+	if !present {
+		return reconcile.Result{}, nil
+	}
+
+	// Make sure the MachineSet has a non-empty selector.
+	msselector, err := metav1.LabelSelectorAsSelector(&instance.Spec.Selector)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if msselector.Empty() {
+		// The cluster-api machinesetcontroller expects every MachineSet to have
+		// its Selector set.
+		log.Info("MachineSet has empty selector, which is unexpected. Will not attempt scaling.")
+		return reconcile.Result{}, nil
+	}
+
+	hostselector, err := actuator.SelectorFromProviderSpec(&instance.Spec.Template.Spec.ProviderSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	hosts := &bmh.BareMetalHostList{}
+	opts := &client.ListOptions{
+		Namespace: instance.Namespace,
+	}
+
+	err = r.List(ctx, hosts, client.UseListOptions(opts))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var count int32
+	for _, host := range hosts.Items {
+		consumer := host.Spec.ConsumerRef
+		if consumer == nil {
+			if hostselector.Matches(labels.Set(host.ObjectMeta.Labels)) {
+				count++
+			}
+		} else {
+			// We will only count this host if it is consumed by a Machine that
+			// is part of the current MachineSet.
+			machine := &machinev1beta1.Machine{}
+			if consumer.Kind != "Machine" || consumer.APIVersion != machinev1beta1.SchemeGroupVersion.String() {
+				// this host is being consumed by something else; don't count it
+				continue
+			}
+
+			// host is being consumed. Let's get the Machine and see if it
+			// matches the current MachineSet.
+			nn := types.NamespacedName{
+				Name:      consumer.Name,
+				Namespace: consumer.Namespace,
+			}
+			err := r.Get(ctx, nn, machine)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					log.Info("Will not scale while BareMetalHost's consuming Machine is not found", "Machine.Name", nn.Name)
+					return reconcile.Result{}, nil
+				}
+				// Error reading the object - requeue the request.
+				return reconcile.Result{}, err
+			}
+			if msselector.Matches(labels.Set(machine.ObjectMeta.Labels)) {
```

Comment:
It seems like it would be safer to check the owner reference. What's the benefit of using the selector instead?

Reply:
The selector is how the machineset controller itself finds matching Machines, so I figured it's safest to use the same mechanism.

Reply:
Interesting, and I agree with using the same mechanism.

Is the selector mutable?

Reply:
I'm not sure if anything enforces it, but it should be immutable. Changing it would make bad things happen. The same mechanism is used by other controllers like ReplicaSets identifying their Pods, Deployments identifying their ReplicaSets, etc.

Reply:
That makes sense. I'll assume for now that if it's not immutable then this won't be any more broken than anything else relying on the matching behavior. :-)